src/tiled/terrainbrush.cpp

```diff
@@ -429,20 +458,20 @@ void TerrainBrush::updateBrush(QPoint cursorPos, const QVector<QPoint> *list)
             mask = 0;
 
             // depending which connections have been set, we update the preferred terrain of the tile accordingly
-            if (y > 0 && checked[i - layerWidth]) {
-                preferredTerrain = (::terrain(newTerrain[i - layerWidth]) << 16) | (preferredTerrain & 0x0000FFFF);
+            if (currentLayer->contains(upPoint) && upperID >= 0 && upperID < numTiles && checked[upperID]) {
```

Comment:
After the `currentLayer->contains` check, you don't need to check `upperID >= 0 && upperID < numTiles` as well, since you already know it will be a valid index. Same applies to the conditions below.